### PR TITLE
Add Canary recipe - wrapper over the SocketCAN API

### PR DESCRIPTION
**Pull request by klimkin**
Specify library name and version:  **canary/1.0.0**

Canary is a thin, C++11 wrapper over the SocketCAN API in Linux, which enables communication over [CAN bus](https://en.wikipedia.org/wiki/CAN_bus) using idioms from the [Boost.ASIO](https://github.com/boostorg/asio) and [ASIO](https://github.com/chriskohlhoff/asio) libraries.

https://github.com/djarek/canary

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/canary/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        cmake.install()
+        rmdir(self, os.path.join(self.package_folder, "lib"))
+
+    def package_id(self):
+        self.info.clear()
+
+    def package_info(self):
+        self.cpp_info.requires = ["boost::headers"]
+        self.cpp_info.set_property("cmake_file_name", "canary")
+        self.cpp_info.set_property("cmake_find_mode", "both")
```
Comment (prince-chrismc) [inline recipes/canary/all/conanfile.py:75]: There's no official find module for this provided by CMake --- please correct me if I am wrong

https://cmake.org/cmake/help/latest/manual/cmake-modules.7.html#find-modules

```suggestion
```

File: recipes/canary/all/test_package/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        self.requires(self.tested_reference_str)
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if can_run(self):
+            bin_path = os.path.join(".", "test_package")
```
Comment (prince-chrismc) [inline recipes/canary/all/test_package/conanfile.py:23]:
```suggestion
            bin_path = os.path.join(self.cpp.build.bindirs[0], "test_package")
```

File: recipes/canary/all/test_package/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+from conan.tools.build import can_run
+from conan.tools.cmake import CMake
+
+
+class TestPackage(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "CMakeToolchain", "CMakeDeps", "VirtualRunEnv"
+    test_type = "explicit"
+
+    def requirements(self):
```
Comment (prince-chrismc) [inline recipes/canary/all/test_package/conanfile.py:13]:
```suggestion
    test_type = "explicit"

    def layout(self):
        cmake_layout(self)

    def requirements(self):
```

File: recipes/canary/all/test_v1_package/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self):
+            bin_path = os.path.join(".", "test_package")
+            self.run(bin_path, run_environment=True)
```
Comment (prince-chrismc) [inline recipes/canary/all/test_v1_package/conanfile.py:18]:
```suggestion
            bin_path = os.path.join("bin", "test_package")
            self.run(bin_path, run_environment=True)
```

File: recipes/canary/config.yml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+versions:
+  "1.0.0":
```
Comment (prince-chrismc) [inline recipes/canary/config.yml:2]:
```suggestion
  "1":
```

**Review by prince-chrismc — COMMENTED** [review]
Over all this looks really good -- just a a bunch of minor comments cleaning up some details

for ex the test package was missing the layout


but super good work

File: recipes/canary/all/conandata.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+sources:
+  "1.0.0":
```
[the remark this answers is not in the document]
Comment (klimkin) [reply recipes/canary/all/conandata.yml:2]: One of the project branches indicates the next release will switch to semver:
https://github.com/djarek/canary/tree/v1.1.0

File: recipes/canary/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    @property
+    def _min_cppstd(self):
+        return 11
+
+    @property
+    def _compilers_minimum_version(self):
+        return {
+            "Visual Studio": "15",
+            "msvc": "14.1",
```
[the remark this answers is not in the document]
Comment (klimkin) [reply recipes/canary/all/conanfile.py]: I have removed compiler checks.

File: recipes/canary/all/conandata.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/canary/all/conandata.yml:2]: @klimkin please, use the official version from the tag. On the next version, we can use semver to the minor. We don't try to fix/modify versions from the upstream, we follow exactly how is it.

[the remark this answers is not in the document]
Comment (klimkin) [reply recipes/canary/all/conandata.yml]: `1` is not a valid version for Conan:
```
$ conan create . canary/1@  --build=missing -pr default
ERROR: Package canary has an invalid version number: '1'. Valid names MUST begin with a letter, number or underscore, have between 2-51 chars, including letters, numbers, underscore, dot and dash
```

[the remark this answers is not in the document]
Comment (prince-chrismc) [reply recipes/canary/all/conandata.yml:2]: Ughs... 🙈 well thats not helpful I also see https://github.com/djarek/canary/issues/20

```suggestion
  "v1":
```

File: recipes/canary/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
+import os
+
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.build import check_min_cppstd
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain
+from conan.tools.files import copy, get, rmdir
+from conan.tools.scm import Version
```
Comment (prince-chrismc) [inline recipes/canary/all/conanfile.py:8]: Linter

```suggestion
```

File: recipes/canary/all/conandata.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (klimkin) [reply recipes/canary/all/conandata.yml:2]: Yes, I have created that issue to inform developers about this PR.

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**